### PR TITLE
implement diagonal method for cupyx.scipy.sparse.dia_matrix

### PR DESCRIPTION
**Pull request by grlee77**
This PR implements the `diagonal` method for the DIA sparse matrix type.

The ``k`` argument (introduced in SciPy 1.0) was also added to the base `spmatrix` class.

### CODE REVIEW

File: cupyx/scipy/sparse/base.py [hunk 1]
```diff
@@ -292,9 +292,9 @@ def count_nonzero(self):
         """Number of non-zero entries, equivalent to"""
         raise NotImplementedError
 
-    def diagonal(self):
+    def diagonal(self, k=0):
         """Returns the main diagonal of the matrix"""
-        return self.tocsr().diagonal()
+        return self.tocsr().diagonal(k=k)
```
Comment (emcastillo) [inline cupyx/scipy/sparse/base.py]: Please, change the docstring to reflect that is the k-th diagonal :)
Thanks!

File: cupyx/scipy/sparse/csr.py [hunk 2]
```diff
@@ -145,10 +145,12 @@ def __rtruediv__(self, other):
     # TODO(unno): Implement argmin
     # TODO(unno): Implement check_format
 
-    def diagonal(self):
+    def diagonal(self, k=0):
         # TODO(unno): Implement diagonal
         raise NotImplementedError
 
+    diagonal.__doc__ = base.spmatrix.diagonal.__doc__
+
```
Comment (emcastillo) [inline cupyx/scipy/sparse/csr.py]: If its not Implemented, the docstring should reflect it too.
Just copying the base one is confusing.

File: cupyx/scipy/sparse/dia.py [hunk 3]
```diff
@@ -180,6 +181,18 @@ def tocsr(self, copy=False):
...
+        rows, cols = self.shape
+        if k <= -rows or k >= cols:
+            raise ValueError("k exceeds matrix dimensions")
+        idx, = cupy.nonzero(self.offsets == k)
+        first_col, last_col = max(0, k), min(rows + k, cols)
+        if idx.size == 0:
+            return cupy.zeros(last_col - first_col, dtype=self.data.dtype)
+        return self.data[idx[0], first_col:last_col]
+
+    diagonal.__doc__ = base.spmatrix.diagonal.__doc__
```
Comment (emcastillo) [inline cupyx/scipy/sparse/dia.py]: Please write the docstring instead of copying it.
Thanks!

**Comment by emcastillo** [comment]
Just some nitpicks :),

Thanks for the contribution!

File: cupyx/scipy/sparse/base.py (hunk 1 above)
Comment (grlee77) [reply cupyx/scipy/sparse/base.py]: done

File: cupyx/scipy/sparse/dia.py (hunk 3 above)
Comment (grlee77) [reply cupyx/scipy/sparse/dia.py]: sure. that was the pattern used in scipy.sparse, but I am fine with duplicating it here

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Please, check CIs

**Comment by grlee77** [comment]
This should be ready to try again on CI. One of the new tests needed `@testing.with_requires('scipy>=1.0')`

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**